tensorflow/compiler/mlir/xla/transforms/xla_hlo_fusion.cc

```diff
@@ -0,0 +1,553 @@
+     return impl_.isEquivalent(lhs, rhs);
+   }
+
+ private:
+  // shape equality propagation based on the shape constrains of
+  // elementwise ops.
+  void PropagateEquality(const SmallVectorImpl<Operation *>& op_list) {
+    bool converged = true;
+    do {
+      converged = true;
+      auto update = [&](Value lhs, Value rhs) {
+        if (!impl_.isEquivalent(lhs, rhs)) {
+          converged = false;
+          impl_.unionSets(lhs, rhs);
+        }
+      };
+      for (Operation* op : op_list) {
+        auto op_fusibility = dyn_cast<InferFusibilityOpInterface>(op);
+        if (!op_fusibility) continue;
+        int numInput = op->getNumOperands();
+        int numOutput = op->getNumResults();
+        // shape equality propagation between inputs.
+        for (int input1 = 0; input1 < numInput; ++input1)
+          for (int input2 = input1+1; input2 < numInput; ++input2)
+            if (op_fusibility.inferInputsShapeEquality(input1, input2))
+              update(op->getOperand(input1), op->getOperand(input2));
+
+        // shape equality propagation between outputs.
+        for (int output1 = 0; output1 < numOutput; ++output1)
+          for (int output2 = output1+1; output2 < numOutput; ++output2)
+            if (op_fusibility.inferOutputsShapeEquality(output1, output2))
+              update(op->getResult(output1), op->getResult(output2));
+
+        // shape equality propagation between input and output.
+        for (int input = 0; input < numInput; ++input)
+          for (int output = 0; output < numOutput; ++output)
+            if (op_fusibility.inferInputOutputShapeEquality(input, output))
+              update(op->getOperand(input), op->getResult(output));
+      }
+    } while (!converged);
+  }
+
+  // a UnionFind set
+  EquivalenceClasses<Value> impl_;
+};
+
+// A fusion planner that can propose a fusion plan for a block of ops.
+// The fusion plan is consisted of a group of fusion patterns.
+//
+// Currently all proposed patterns followed xla kLoop/kInput like fusion
+// templates while are adapted to the fully dynamic shape world.
+//
+// kLoop fusion template satifies:
+//   - all ops in the fusion pattern are element-wise.
+//   - all the shapes of outputs of fusion pattern are same, and thus can
+//     fit into a same parallel loop.
+//
+// kInput fusion template satifies:
+//   - any op in the fusion pattern is either element-wise or a reduction.
+//   - if a op is a reduction, its output cannot be consumered by other
+//     ops in the same fusion pattern.
+//   - all the effective shapes of outputs of fusion pattern are same.
+//     - For element-wise op, its effective shape is its output shape.
+//     - For reduction op, its effective shape is its operand shape.
+class FusionPlanner {
+ public:
+  explicit FusionPlanner(const SmallVectorImpl<Operation *>& op_list)
+   : op_list_(op_list), shape_analysis_(op_list) {
+    BuildNodeMap();
+  }
+
+  // Return a fusion plan if success, otherwise none.
```

Comment:
I don't see the none part, is this just to enable future error reporting ?

Reply:
Yes, it's reserved for the future.